python/hsfs/core/hudi_engine.py

```diff
@@ -0,0 +1,195 @@
+
+    def save_hudi_fg(self, dataset, save_mode, operation, write_options):
+        fg_commit = self._write_hudi_dataset(
+            dataset, save_mode, operation, write_options
+        )
+
+        return self._feature_group_api.commit(self._feature_group, fg_commit)
+
+    def register_temporary_table(self, alias, start_timestamp, end_timestamp):
+        self._spark_session.conf.set("spark.sql.hive.convertMetastoreParquet", "false")
+
+        hudi_options = self._setup_hudi_read_opts(start_timestamp, end_timestamp)
+        self._spark_session.read.format(self.HUDI_SPARK_FORMAT).options(
+            **hudi_options
+        ).load(self._base_path).registerTempTable(alias)
+
+    def _write_hudi_dataset(self, dataset, save_mode, operation, write_options):
+        hudi_options = self._setup_hudi_write_opts(operation)
+        if bool(write_options):
+            hudi_options.update(write_options)
+
+        dataset.write.format(HudiEngine.HUDI_SPARK_FORMAT).options(**hudi_options).mode(
+            save_mode
+        ).save(self._base_path)
+
+        feature_group_commit = self._get_last_commit_metadata(
+            self._spark_context, self._base_path
+        )
+
+        return feature_group_commit
+
+    def _setup_hudi_write_opts(self, operation):
+        _jdbc_url = self._get_conn_str()
+        hudi_options = {
+            self.HUDI_PRECOMBINE_FIELD: self._precombine_key,
+            self.HUDI_RECORD_KEY: self._primary_key,
+            self.HUDI_PARTITION_FIELD: self._partition_key,
+            self.HUDI_TABLE_NAME: self._table_name,
+            self.HIVE_PARTITION_EXTRACTOR_CLASS_OPT_KEY: self.DEFAULT_HIVE_PARTITION_EXTRACTOR_CLASS_OPT_VAL,
+            self.HUDI_HIVE_SYNC_ENABLE: "true",
+            self.HUDI_HIVE_SYNC_TABLE: self._table_name,
+            self.HUDI_HIVE_SYNC_JDBC_URL: _jdbc_url,
+            self.HUDI_HIVE_SYNC_DB: self._feature_store_name,
+            self.HUDI_TABLE_OPERATION: operation,
+        }
+
+        return hudi_options
+
+    def _setup_hudi_read_opts(self, start_timestamp, end_timestamp):
+        _hudi_commit_start_time = self._timestamp2hudiformat(start_timestamp)
+        _hudi_commit_end_time = self._timestamp2hudiformat(end_timestamp)
+
+        hudi_options = {
+            self.HUDI_QUERY_TYPE_OPT_KEY: self.HUDI_QUERY_TYPE_INCREMENTAL_OPT_VAL,
+            self.HUDI_BEGIN_INSTANTTIME_OPT_KEY: _hudi_commit_start_time,
+            self.HUDI_END_INSTANTTIME_OPT_KEY: _hudi_commit_end_time,
+        }
+
+        return hudi_options
+
+    def _get_last_commit_metadata(self, spark_context, base_path):
+        hopsfs_conf = spark_context._jvm.org.apache.hadoop.fs.FileSystem.get(
+            spark_context._jsc.hadoopConfiguration()
+        )
+        commit_timeline = spark_context._jvm.org.apache.hudi.HoodieDataSourceHelpers.allCompletedCommitsCompactions(
+            hopsfs_conf, base_path
+        )
+
+        commits2return = commit_timeline.getInstantDetails(
```

Comment:
Let's try to be consistent and use snake case everywhere.

Reply:
done